phylo/acceleration.go

```diff
@@ -0,0 +1,339 @@
+	return reachedEnd
+}
+
+//calculateWeights is a helper function of the distanceAndWeights functions, produces the weight constant for each distance.
+func calculateWeights(d *AccelDistancesAndWeights, zeroDistanceWeightConstant float64) {
+	d.WhumChimp = calculateWeight(d.DhumChimp, zeroDistanceWeightConstant)
+	d.WhumGor = calculateWeight(d.DhumGor, zeroDistanceWeightConstant)
+	d.WhumOrang = calculateWeight(d.DhumOrang, zeroDistanceWeightConstant)
+	d.WchimpGor = calculateWeight(d.DchimpGor, zeroDistanceWeightConstant)
+	d.WchimpOrang = calculateWeight(d.DchimpOrang, zeroDistanceWeightConstant)
+	d.WgorOrang = calculateWeight(d.DgorOrang, zeroDistanceWeightConstant)
+}
+
+//calculateWeight is a helper function of calculateWeights, and performs each individual weight calculation.
+func calculateWeight(d float64, zeroDistanceWeightConstant float64) float64 {
+	if d == 0 {
+		return zeroDistanceWeightConstant
+	}
+	return 1.0 / math.Pow(d, 2)
+}
+
+//IsUngappedColumn determines if an alignment column is comprised of bases (not gaps) for each species.
+func IsUngappedColumn(records []fasta.Fasta, index int) bool {
+	for i := range records {
+		if !isUngappedBase(records[i].Seq[index]) {
+			return false
+		}
+	}
+	return true
+}
+
+//isUngappedBase is a helper function of isUngappedColumn. True if a dna.Base is a base, not an N, gap, or dot.
+func isUngappedBase(b dna.Base) bool {
+	if b == dna.A || b == dna.T || b == dna.C || b == dna.G {
+		return true
+	}
+	if b == dna.LowerA || b == dna.LowerC || b == dna.LowerG || b == dna.LowerT {
+		return true
+	}
+	return false
+}
+
+//A helper function of BranchLengthsAlternatingLeastSquares. Reduce the four species tree to the subtree containing species 0, 1, and the ancestor of 2/3.
+func pruneLeft(d AccelDistancesAndWeights, b AccelBranchLengths, sub *AccelSubTreeLeft, ZeroDistanceWeightConstant float64) {
+	sub.DhumChimp = d.DhumChimp
+	sub.DhumHga = (d.WhumGor*(d.DhumGor-b.BhgaGor) + d.WhumOrang*(d.DhumOrang-b.BhgaOrang)) / (d.WhumGor + d.WhumOrang)
+	sub.DchimpHga = (d.WchimpGor*(d.DchimpGor-b.BhgaGor) + d.WchimpOrang*(d.DchimpOrang-b.BhgaOrang)) / (d.WchimpGor + d.WchimpOrang)
+
+	sub.WhumChimp = calculateWeight(sub.DhumChimp, ZeroDistanceWeightConstant)
+	sub.WhumHga = calculateWeight(sub.DhumHga, ZeroDistanceWeightConstant)
+	sub.WchimpHga = calculateWeight(sub.DchimpHga, ZeroDistanceWeightConstant)
+}
+
+//A helper function of BranchLengthsAlternatingLeastSquares. Reduce the four species tree to the subtree containing 2, 3, and the ancestor of 0/1.
+func pruneRight(d AccelDistancesAndWeights, b AccelBranchLengths, sub *AccelSubTreeRight, ZeroDistanceWeightConstant float64) {
+	sub.DgorOrang = d.DgorOrang
+	sub.DhcaGor = (d.WhumGor*(d.DhumGor-b.BhumHca) + d.WchimpGor*(d.DchimpGor-b.BchimpHca)) / (d.WhumGor + d.WchimpGor)
+	sub.DhcaOrang = (d.WhumOrang*(d.DhumOrang-b.BhumHca) + d.WchimpOrang*(d.DchimpOrang-b.BchimpHca)) / (d.WhumOrang + d.WchimpOrang)
+
+	sub.WgorOrang = d.DgorOrang
```

Comment:
I have not looked back at the paper, but I expected the pruneRight to be more similar to pruneLeft, but it looks pretty different with no call to the calculateWeight function.

Reply:
great catch!